blog/_posts/2017-02-25-collections-rework.md

```diff
@@ -0,0 +1,218 @@
+
+As shown above, the new design does not have an implicit evidence parameter for `map`, so you are assured that you always get the same implementation no matter if you statically see your `BitSet` as a `BitSet` or a `Set[Int]`. This "same implementation" in the case of `BitSet` means `Set` though. A collection implementation is free to pick any collection type to build (between what its supertype promises and what it implements itself, of course) as long as it is *unconstrained*. In order to call `BitSet.map` and get another constrained `BitSet` out of it, we need to overload the `map` method:
+
+```scala
+class BitSet extends Set[Int] {
+  // inherited:
+  //def map[B](f: Int => B): Set[B]
+
+  def map(f: Int => Int): BitSet
+}
+```
+
+Thanks to [an improvement to type inference](https://github.com/scala/scala/pull/5307) it is possible to call the overloaded method with a lambda without explicit type annotations in Scala 2.12:
+
+```scala
+val s: BitSet = ...
+
+// Scala 2.11 and earlier would have required:
+val c1 = s.map((i: Int) => i+1)
+
+// This works in 2.12+:
+val c2 = s.map(i => i+1)
+```
+
+While `BitSet` is a rather esoteric collection type, the same principle of implicit constraints on element types `T` applies to other collection types as well:
+
+- BitSet: `T <:< Int` ("must be an Int")
+- All `Map` types: `T <:< (_, _)` ("must be a Tuple2")
+- All sorted collections (like `TreeSet`): `Ordering[T]` ("must have an Ordering")
+- String (not a collection type per se but it gets many collection methods as extension methods): `T <:< Char` ("must be a Char")
+
+This covers most uses of `CanBuildFrom` in a much simpler way. The most important use cases that cannot be supported by the new design are `collection.breakOut` (for building a different collection type directly without an extra conversion step) and `to` (for converting to a different collection type for which a `CanBuildFrom` is available).
+
+## FromIterable
+
+The `to` method still exists in the new design but it is only a decorator around `FromIterable`, the basic abstraction that is implemented by every unconstrained collection type's companion object:
+
+```scala
+trait IterableOps[+A] extends Any {
+  def to[C[X] <: Iterable[X]](fi: FromIterable[C]): C[A @uncheckedVariance] =
+    fi.fromIterable(coll)
+  ...
+}
+
+trait FromIterable[+C[X] <: Iterable[X]] {
+  def fromIterable[B](it: Iterable[B]): C[B]
+}
+```
+
+Note that the `fi` parameter is not implicit (like the `CanBuildFrom` in the current library), so you now call `to` with a value and have the type inferred instead of calling it with a type and having the value filled in by implicit lookup:
+
+```scala
+val s: Set[Int] = ...
+val v = s.to(Vector) // this used to be to[Vector]
+```
+
+This has the advantage that we can overload `to` to cover maps (and maybe also other constrained collection types) which is not possible in the current design (where `to` only works for unary type constructors).
+
```

Comment:
It could be part of a migration library. If we want to provide source compatibility for building on 2.12 and 2.13 we probably need extra libraries for both versions.